README.md
# snompnpd

snom Plug and Play daemon.

snompnpd is a deamon that handle Plug&Play provisioning requests sent by snom phones.

### Features

* multiple matching rules, you can distinguish between phone models, mac address, firmware, ecc..
* multiple network interface support
* custom response (IGNORE or DENY specific requests)
* execution of external commands 

### The process

The PnP process are described [here](http://wiki.snom.com/Features/Auto_Provisioning/PnP).

### Installation

#### Configuration

* you need to modify some variables in CONFIG file:

  * **DESTDIR** defines the installation directory
  * **PYTHON_LIB** defines the directory where all needed python modules will be installed
  * **INITDIR** defines the directory where the init script will be placed
  * **DAEMONDIR** defines where the main script will be placed
  * **CONFDIR** defines the configuration directory
  * **USER** and **GROUP** manages the ownership
  * **PIDFILE** defines where the daemon PID must be saved (some distro requires it in init.d script)
  * **LOGFILE** the logfile
  * **TARGET** defines the distributrion where the daemon is installed, supported values: *rh* and *deb*, this affects the init script

#### Building it

run the comannd

    make install

## Configuring the daemon

The installation process will install a sample configuration file named *snompnpd.conf* in the directory defined by **CONFDIR** into *CONFIG* file.

The config file defines all needed parameters inside the **main** section.
**rule_XX* sections contain matching rules.

#### The main section

**local_ip**: this parameter defines the ip address used in answering the phone requests. Differents values are allowed:

  * *default*: the daemon will autodiscover the local ip reaching an host (snom.com)
  * *xxx.xxx.xxx.xxx*: an IP address: you can insert a local IP address to use as a source address
  * *eth0*: a network interface name: the daemon will use the IP address associated to an interface
  * *0.0.0.0*: using this value the daemon doesn't binds to a specific IP address
  
**unicast_port**: this defines the source port for generated responses (ignored in case of *local_ip=0.0.0.0*)  

**pid_file**: a file where the PID will be stored

**log_file**: here you can insert a file path or a syslog server definition following this syntax: syslog:IP_ADDRESS:PORT:FACILITY

Eg.: *syslog:172.16.18.99:514:local7* will send all messages to the remote syslog on 172.16.18.99, using the 514 UDP port and the local7 facility.

**debug**: here you can insert True or False, use True only during debug purpose

**daemon**: another boolean value, using False the program will start in foreground, useful for debugging purpose

#### Matching rules

Every matching rule is defined by a section named *rule_XX* where XX defines the ordering number. Every rule must contains a *regex* and *subst* statement.

When a SUBSCRIBE request comes the daemon starts looping inside the rules, for each rule the regular expression defined by *regex* is evaluaed against the full SUBSCRIBE (headers and body too). If the pattern is found, a substitution defined by *subst* is applied and the resulting string is used as body in NOTIFY sent by the daemon to the phone.

When a pattern rule is found the evaluation loop stops, so the first matching rule wins.

In subst statement you can use the special placeholder *${local_ip}* this placeholder will be substituted with the local ip address (discovered following same rules of *local_ip* configuration paramether).

For more informations about python regex see: 

* [http://docs.python.org/2/library/re.html](http://docs.python.org/2/library/re.html)
* [http://docs.python.org/2/howto/regex.html](http://docs.python.org/2/howto/regex.html)

##### Denying requests

When subst statement starts with DENY prefix the daemon doesn't send the NOTIFY and reply with a SIP Response code and message extracted from the DENY statement 

Eg.:
 
    subst=DENY:401:Unauthorized you're not allowed

replies with a "SIP/2.0 401 Unauthorized you're not allowed" message

##### Ignoring requests

    subst=DISCARD
    
When subst statement is DISCARD the daemon ignores the matched request: This can be useful in some installation with multiple PnP deamons running in the same network

#### Executing external commands

A rule can contains also a *command* statement that is meant for executing a command in case of a rule matching. The command is executed in a new process, with these environment variables:

* **RULE_IDX**: a number showing the matched rule index
* **REGEX**: the content in matched rule *regex* statement
* **SUBST**: the content in matched rule *subst* statement
* **REGEX_GROUP_X**: a variable for each matching group contained in *regex* statement, the group index is defined by the **X** value.

**NB: Be careful in evauating REGEX_GROUP_ variable in order to avoid any shell injection.** 

## Running the daemon

You can start the daemon trough the init.d script:

    /etc/init.d/snompnpd start

Or manually (after configuring the PYTHNPATH variable according with *PYTHON_LIB* defined in *CONFIG* file):

    export PYTHONPATH=$PYHONPATH:"/opt/snompnpd/lib/python"
    /opt/snompnpd/usr/sbin/snompnpd -s /etc/snompnpd.conf 

## Examples

#### Matching all requests rule

Matching rule:

    regex = (.*\n)*
    subst = http://172.16.16.5
 
SUBSCRIBE Request sent from the phone to **sip.mcast.net**:

    SUBSCRIBE sip:MAC%3a00135E874B49@fake SIP/2.0
    Content-Length: 0
    Via: SIP/2.0/UDP 172.16.18.90:5060;rport
    From: <sip:MAC%3a00135E874B49>;tag=658512961
    Expires: 0
    Accept: application/url
    To: <sip:MAC%3a00135E874B49>
    Contact: <sip:172.16.18.90:5060>
    CSeq: 1 SUBSCRIBE
    Call-ID: 1930770594@ciccio
    Event: ua-profile;profile-type="device";vendor="snom";model="snom720";version="7.1.19"

Messages sent by the daemon to the phone:

SIP 200 OK:

    SIP/2.0 200 OK
    Content-Length: 0
    Via: SIP/2.0/UDP 172.16.18.90:5060;rport
    From: <sip:MAC%3a00135E874B49>;tag=658512961
    Expires: 0
    To: <sip:MAC%3a00135E874B49>
    Contact: <sip:172.16.18.90:53465;transport=udp;handler=dum>
    CSeq: 1 SUBSCRIBE
    Call-ID: 1930770594@ciccio

SIP NOTIFY:

    NOTIFY sip:172.16.18.90:5060 SIP/2.0
    Content-Length: 18
    Via: SIP/2.0/UDP 172.16.18.90:5060;rport
    From: <sip:MAC%3a00135E874B49>;tag=658512961
    Subscription-State: terminated;reason=timeout
    To: <sip:MAC%3a00135E874B49>
    Contact: <sip:172.16.18.90:53465;transport=udp;handler=dum>
    CSeq: 3 NOTIFY
    Max-Forwards: 20
    Call-ID: 1930770594@ciccio
    Content-Type: application/url
    Event: ua-profile;profile-type="device";vendor="snom720";model="snom720";version="7.1.19"
    
    http://172.16.18.5

#### Denying a specific phone model and executing an external command

Matching rule:

    ; DENY snom300 models only
    regex = ^SUBSCRIBE\s.*MAC%3a(.*)@.*\n(.*\n)*Event:\s.*model="snom300";.*\n(.*\n)*
    subst = DENY:401:Unauthorized: go away
    command = /opt/snompnpd/logdeny.sh

The script **/opt/snompnpd/logdeny.sh** (don't forget to make it executable):

	#!/bin/bash
	
	DATE=$(date)
	LOGFILE=/tmp/unwanted-snom300.log
	
	# Regex Rule in config file:
	# regex = ^SUBSCRIBE\s.*MAC%3a(.*)@.*\n(.*\n)*Event:\s.*model="snom300";.*\n(.*\n)*
	#                             ^  ^     ^    ^                               ^    ^
	#                             |  |     |    |                               |    |
	#             $REGEX_GROUP_1 -+--+     |    |                               |    |
	#             $REGEX_GROUP_2 ----------+----+                               |    |
	#             $REGEX_GROUP_2 -----------------------------------------------+----+
	
	PHONE_MAC=$REGEX_GROUP_1
	
	echo "$DATE: request received by $PHONE_MAC" >> $LOGFILE
	
	# DEBUG ONLY: save the environment
	# echo "------START ENV------" >> /tmp/log.txt
	# env >> /tmp/log.txt
	# echo "------STOP ENV------" >> /tmp/log.txt
	# echo "ARGS: $@" >> /tmp/log.txt

SUBSCRIBE Request sent by the phone to **sip.mcast.net**:

    SUBSCRIBE sip:MAC%3a00135E874B49@fake SIP/2.0
    Content-Length: 0
    Via: SIP/2.0/UDP 172.16.18.90:5060;rport
    From: <sip:MAC%3a00135E874B49>;tag=658512961
    Expires: 0
    Accept: application/url
    To: <sip:MAC%3a00135E874B49>
    Contact: <sip:172.16.18.90:5060>
    CSeq: 1 SUBSCRIBE
    Call-ID: 1930770594@ciccio
    Event: ua-profile;profile-type="device";vendor="snom";model="snom300";version="7.1.19"

Messages sent by the daemon to the phone:

    SIP/2.0 401 Unauthorized: go away
    Content-Length: 0
    Via: SIP/2.0/UDP 172.16.18.90:5060;rport
    From: <sip:MAC%3a00135E874B49>;tag=658512961
    Expires: 0
    To: <sip:MAC%3a00135E874B49>
    Contact: <sip:172.16.18.90:55365;transport=udp;handler=dum>
    CSeq: 1 SUBSCRIBE
    Call-ID: 1930770594@ciccio

Content of */tmp/unwanted-snom300.log*:

    Wed Oct 30 12:40:02 CET 2013: request received by 00135E874B49

## Debugging

You can configure the *debug=True* statement in config file and in order to log all SIP messages and daemon activities. Using *daemon=False* the process will not detached, without defining a *log_file* all log messages will be printed on stdout.

## Multiple networks

In case your host is using more than one network interface you should make sure the multicast address *sip.mcast.net* is boud to the proper interface. By default is bound to the default network device, in case you want to change it you shoud create a static route:

*Tho following command will bind the sip.mcast.net address to the network device eth1:*

    sudo ip route add 224.0.1.75 dev eth1


## Usage

This software is released for didactical and debugging purposes. You're free to use it at your own risk. You can modify and redistribute this program under the [LGPLv3](http://www.gnu.org/licenses/lgpl-3.0.txt) license terms.

[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/pbertera/snompnpd/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



src/changelog
snompnpd (0.0.5-1) stable; urgency=low

  * fixed small bug in send() udp function

 -- Pietro Bertera <pietro.bertera@snom.com> Tue, 06 Mar 2018 15:41:32 +0100

snompnpd (0.0.5-1) stable; urgency=low

  * binding to multiple address support
  * introduced command execution on matched requests

 -- Pietro Bertera <pietro.bertera@snom.com>  Tue, 29 Oct 2013 10:02:31 +0100

snompnpd (0.0.4-1) stable; urgency=low

  * Improved Makefile and installation process
  * added snomprovisioning python packages with daemon and sip modules

 -- Pietro Bertera <pietro.bertera@snom.com>  Tue, 27 Mar 2013 19:22:21 +0100

snompnpd (0.0.3-1) stable; urgency=low

  * added "DISCARD" rule

 -- Pietro Bertera <pietro.bertera@snom.com>  Tue, 27 Mar 2013 19:22:21 +0100

snompnpd (0.0.2-1) stable; urgency=low

  * added ${local_ip} placeholder support in subst definition
  * added validation of ip address in local_ip
  * added ip address discovery via net device name

 -- Pietro Bertera <pietro.bertera@snom.com>  Tue, 19 Mar 2013 19:17:53 +0100

snompnpd (0.0.1-1) stable; urgency=low

  * Initial packaging

 -- Pietro Bertera <pietro.bertera@snom.com>  Mon, 18 Feb 2013 15:24:11 +0100


src/snompnpd.py
#!/usr/bin/python
# vi:si:et:sw=4:sts=4:ts=4
# -*- coding: UTF-8 -*-
# -*- Mode: Python -*-

import socket
import struct
import shlex
import re
import sys
import select
import ConfigParser
import os
import signal
import logging
import logging.handlers
import cStringIO
import subprocess
from string import Template

from snomprovisioning import sip
from snomprovisioning import daemon

# a workaround to avoid zoombie child in external commands
import signal
signal.signal(signal.SIGCHLD, signal.SIG_IGN)

version = "__VERSION__"

logging.handlers.raiseExceptions = True

logger = logging.getLogger(sys.argv[0])
log_handler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
log_handler.setFormatter(formatter)
logger.addHandler(log_handler)
logger.setLevel(logging.INFO)

class Error:
    def __init__(self, message=None):
        if message:
            logger.error(message)
        pass

class UnsupportedSIPVersion(Error): pass
class UnsupportedSIPTransport(Error): pass
class SendDataError(Error): pass
class SystemError(Error): pass
class ConfigError(Error): pass

class Debug:
    def __init__(self, message=None):
        if message:
            logger.debug(message)
        pass

class DebugMessage(Debug): pass

class Info:
    def __init__(self, message=None):
        if message:
            logger.info(message)
        pass

class CreateSocketError(Error): pass

class InfoMessage(Info): pass

class SocketOptionError(Error): pass
class BindSocketError(Error): pass
class InvalidRegexError(Error): pass

class MCastServer:
    def __init__(self, local_ip,
            regex=[r'^SUBSCRIBE\s.*MAC%3a(.*)@.*\n(.*\n)*Event:\s.*model="(.*)";.*\n(.*\n)*'],
            subst=[r'http://provisioning.snom.com/\3/\3.php?mac=\1'], run_commands=[], mcast=True, mport=5060, uport=5060):

        self.mcast = mcast
        self.port = mport #listen port for multicast requests
        self.uni_port = uport #source port for unicast requests

        self.local_ip = local_ip
        if self.mcast:
            self.listen_addr = "224.0.1.75"
        else:
            self.listen_addr = local_ip

        DebugMessage("Local unicast IP: %s" % self.local_ip)
        DebugMessage("Multicast listening IP: %s" % self.listen_addr)


        self.recvsocket = self._create_mcast_socket()
        self.regex = []

        for x in regex:
            try:
                self.regex.append(re.compile(x, re.MULTILINE))
            except Exception:
                raise InvalidRegexError(x)

        self.subst = subst
        self.run_commands = run_commands

    def listen(self):
        # Sockets from which we expect to read
        inputs = [self.recvsocket]
        # Sockets to which we expect to write
        outputs = [ ]

        DebugMessage("Starting listening loop")

        while inputs:

            readable, writable, exceptional = select.select(inputs, outputs, inputs)
            # Handle inputs
            for s in readable:
                # is SUBSCRIBE
                if s is self.recvsocket:
                    subscription = self.recvsocket.recv(10240)
                    try:
                        request = sip.Request(subscription)
                        if request.method != "SUBSCRIBE":
                            DebugMessage("Received a non SUBSCIBE: %s" % request.method)
                            continue
                    except sip.SipUnpackError:
                        # NOT a SUBSCRIBE
                        continue

                    matched = self.check_request(request)
                    if matched:
                        response = sip.Response()
                        if self.subst[self.matched_rule_index].startswith("DENY:"):
                            response.reason = ":".join(self.subst[self.matched_rule_index].split(":")[2:])
                            response.status = self.subst[self.matched_rule_index].split(":")[1]
                        elif self.subst[self.matched_rule_index] == "DISCARD":
                            DebugMessage("Ignoring request due to 'DISCARD' rule")
                            continue
                        response.headers['from'] = request.headers['from']
                        response.headers['to'] = request.headers['to']
                        response.headers['call-id'] = request.headers['call-id']
                        response.headers['cseq'] = request.headers['cseq']
                        response.headers['expires'] = 0
                        response.headers['via'] = request.headers['via']
                        response.headers['contact'] = "<sip:%s:%d;transport=udp;handler=dum>" % (self.local_ip, self.uni_port)
                        response.headers['content-length'] = 0
                        #Regexp parsing via Header: SIP/2.0/UDP 172.16.18.90:5060;rport
                        p = re.compile(r'SIP/(.*)/(.*)\s(.*):([0-9]*);*')
                        m = p.search(request.headers['via'])

                        if m:
                            version = m.group(1)
                            transport = m.group(2)
                            if version != "2.0":
                                UnsupportedSIPVersion("Unsupported SIP version in Via: header: %s" % version)
                                continue
                            phone_ip = m.group(3)
                            phone_port = m.group(4)
                            #phone_port = "5060"
                        else:
                            Error("Wrong Via: header")

                        if transport.upper() == "UDP":
                            DebugMessage("Creating send socket")
                            try:
                                self.sendsock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
                                if self.local_ip == "0.0.0.0":
                                    self.sendsock.connect((phone_ip, int(phone_port)))
                                    ip = self.sendsock.getsockname()[0]
                                    port = self.sendsock.getsockname()[1]
                                    DebugMessage("Sending trough local IP: %s" % ip)
                                    DebugMessage("Using local Port: %s" % port)
                                    response.headers['contact'] = "<sip:%s:%d;transport=udp;handler=dum>" % (ip, port)
                            except Exception, e:
                                CreateSocketError("Cannot create socket: %s" % e)
                                continue
                            try:
                                self.sendsock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                                self.sendsock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
                            except AttributeError:
                                DebugMessage("Cannot set SO_REUSEADDR or SO_REUSEPORT")
                        else:
                            UnsupportedSIPTransport("Unsupported Transport in Via: header")
                            continue

                        if self.local_ip != "0.0.0.0":
                            try:
                                DebugMessage("Binding to local ip:port %s:%s" % (self.local_ip, self.uni_port))
                                self.sendsock.bind((self.local_ip, self.uni_port))
                            except Exception, e:
                                SendDataError("Cannot bind socket to %s:%d: %s" % (self.local_ip, self.uni_port, e))

                        # sent the OK (or 404 if DENY)
                        try:
                            DebugMessage("Sending response to %s:%s : \n%s" % (phone_ip, phone_port, str(response)))
                            self.sendsock.sendto(str(response), (phone_ip, int(phone_port)))
                        except Exception, e:
                            SendDataError("Cannot send OK/DENY response to %s:%s: %s" % (phone_ip, phone_port, e))
                            self.sendsock.close()
                            continue

                        # if DENY doesn't send the NOTIFY
                        if self.subst[self.matched_rule_index].startswith("DENY:"):
                            self.sendsock.close()
                            continue

                        # create the NOTIFY message
                        cseq = int(request.headers['cseq'].split()[0]) + 2

                        DebugMessage("Forging NOTIFY")
                        notify = sip.Request()
                        notify.method = "NOTIFY"
                        notify.version ="2.0"
                        # Build the URI using IP and Port,
                        notify.uri = "sip:%s:%s" % (phone_ip, phone_port)
                        if request.headers.has_key("contact"):
                            uri_regex = re.compile("sip:([^@]*)@([^>$]*)")
                            addr_regex = re.compile("sip:([^ ;>$]*)")
                            md = uri_regex.search(request.headers["contact"])
                            if md: # Contact: contains a SIP URI
                                notify.uri = "sip:%s@%s" % (md.group(1), md.group(2))
                                DebugMessage("Found URI in SUBSCRIBE Contact")
                            else:
                                md = addr_regex.search(request.headers["contact"])
                                if md: # Contact: contains an address
                                    notify.uri = "sip:" + md.group(1)
                                    DebugMessage("Found Address in SUBSCRIBE Contact")
                                else: # Malformed Contact?
                                    InfoMessage("Found Nothing in SUBSCRIBE Contact")
                        else:
                            InfoMessage("Contact header not found in SUBSCRIBE")

                        notify.headers['via'] = request.headers['via']
                        notify.headers['max-forwards'] = 20
                        notify.headers['to'] = request.headers['to']
                        notify.headers['from'] = request.headers['from']
                        notify.headers['call-id'] = request.headers['call-id']
                        notify.headers['cseq'] = "%d NOTIFY" % cseq
                        notify.headers['content-type'] = "application/url"
                        notify.headers['subscription-state'] = "terminated;reason=timeout"
                        notify.headers['event'] = 'ua-profile;profile-type="device";vendor="OEM";model="OEM";version="7.1.19"'

                        try:
                            if self.local_ip == "0.0.0.0":
                                self.sendsock.connect((phone_ip, int(phone_port)))
                                ip = self.sendsock.getsockname()[0]
                                port = self.sendsock.getsockname()[1]
                                DebugMessage("Sending trough local IP: %s" % ip)
                                DebugMessage("Using local Port: %s" % port)
                                notify.headers['contact'] = "<sip:%s:%d;transport=udp;handler=dum>" % (ip, port)
                                subst = Template(self.subst[self.matched_rule_index]).substitute(local_ip=ip)
                                body = self.regex[self.matched_rule_index].sub(subst, str(request))
                            else:
                                notify.headers['contact'] = "<sip:%s:%d;transport=udp;handler=dum>" % (self.local_ip, self.uni_port)
                                subst = Template(self.subst[self.matched_rule_index]).substitute(local_ip=local_ip)
                                body = self.regex[self.matched_rule_index].sub(subst, str(request))

                            notify.body = body
                            notify.headers['content-length'] = "%d" % len(body)
                            DebugMessage("Sending NOTIFY to %s:%s : \n%s" % (phone_ip, phone_port, str(notify)))
                            self.sendsock.sendto(str(notify), (phone_ip, int(phone_port)))
                        except Exception, e:
                            SendDataError("Cannot send NOTIFY request to %s:%s: %s" % (phone_ip, phone_port, e))
                        self.sendsock.close()

    def check_request(self, request):
        i = 0
        self.matched_rule_index = 0
        res = False
        self.matched_groups = ()
        DebugMessage("Received request")
        DebugMessage("Search:")
        DebugMessage(request)
        for r in self.regex:
            DebugMessage("Evaluating %d rule" % i)
            DebugMessage("Pattern: %s" % r.pattern)
            DebugMessage("Subst: %s" % self.subst[i])
            try:
                #if self.local_ip == "0.0.0.0":
                    #DebugMessage("Applying temporary local_ip = 0.0.0.0 substitution")
                    #tmp_sub = Template(self.subst[i].substitute(local_ip=local_ip))
                #res = r.sub(self.subst[i], str(request))
                res = r.search(str(request))
            except Exception, e:
                InvalidRegexError("Error applying the regex/subst %d: %s" % (i, e))
                i = i + 1
                continue
            if res == None:
                i = i + 1
                continue
            else:
                InfoMessage("snom PnP matching request found: rule number: %d" % i)
                m = r.search(str(request))
                if m:
                    self.matched_groups = m.groups()
                #DebugMessage("Result: %s" % res)
                self.matched_rule_index = i
                n = 1
                command_env = {}
                for g in self.matched_groups:
                    command_env["REGEX_GROUP_%d" % n] = g
                    command_env["REGEX"] = r.pattern
                    command_env["SUBST"] = self.subst[i]
                    command_env["RULE_IDX"] = "%d" % i
                    n = n + 1
                if self.run_commands[self.matched_rule_index]:
                    try:
                        DebugMessage("External command env: %s" % command_env)
                        DebugMessage("Execuding external command: %s" % self.run_commands[self.matched_rule_index])
                        command_pid = subprocess.Popen(shlex.split(self.run_commands[self.matched_rule_index]), env=command_env).pid
                        InfoMessage("Spawned command '%s' with pid %d" % (self.run_commands[self.matched_rule_index], command_pid))
                    except Exception, e:
                        InfoMessage("Error executing command: %s" % e)
                return True
            i = i + 1
        DebugMessage("No Match Found")
        return False

    def _create_mcast_socket(self):
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
            sock.setblocking(0)
        except Exception, e:
            raise CreateSocketError("Cannot create socket: %s" % e)
        try:
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        except AttributeError:
            pass
        try:
            sock.bind((socket.gethostbyname(self.listen_addr), self.port))
        except Exception, e:
            raise BindSocketError("Cannot bind socket to %s:%d: %s" % (self.listen_addr, self.port, e))
        if self.mcast:
            mreq = struct.pack('4sl', socket.inet_aton(socket.gethostbyname(self.listen_addr)), socket.INADDR_ANY)
            try:
                sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)
            except socket.error, e:
                raise SocketOptionError("Cannont configure multicast: %s" % e)
        return sock


def get_ip_address(host="snom.com"):
    # This is a simple hack to find our IP address
    # AFAIK this is the only platform-independent way to obtain the address
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect((host, 80))
    ip = s.getsockname()[0]
    s.close()
    return ip

def is_valid_ipv4_address(address):
    try:
        addr= socket.inet_pton(socket.AF_INET, address)
    except AttributeError: # no inet_pton here, sorry
        try:
            addr= socket.inet_aton(address)
        except socket.error:
            return False
        return address.count('.') == 3 #invalidate shortened address (like 127.1)
    except socket.error: # not a valid address
        return False
    return True

def get_ip_address_ifname(iface):
    import fcntl
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    return socket.inet_ntoa(fcntl.ioctl(
        s.fileno(),
        0x8915,  # SIOCGIFADDR
        struct.pack('256s', iface[:15])
    )[20:24])

if __name__ == '__main__':

    def usage():
        print "\nUsage: %s [options] <config-file>" % sys.argv[0]
        print "\n\tOptions:"
        print "\t\t-c <config-file>\t\tRead a SIP request from stdin and print the response, useful for testing rules"
        print "\t\t-s <config-file>\t\tStart the program"
        sys.exit()

    def signal_handler(signal, frame):
        InfoMessage('Killed by SIGTERM. Goodbye.')
        sys.exit(0)

    signal.signal(signal.SIGTERM, signal_handler)


    # TODO: improve command line handling
    if len(sys.argv) < 3:
        print "\nERROR: missing args."
        usage()

    if sys.argv[1] not in ["-c", "-s"]:
        print "\nERROR: wrong arg."
        usage()

    config = ConfigParser.RawConfigParser()
    config.read(sys.argv[2])

    try:
        debug = config.get('main', 'debug')
        if debug.upper() == 'TRUE':
            logger.setLevel(logging.DEBUG)
            DebugMessage("Log level: DEBUG")

    except ConfigParser.NoOptionError:
        pass

    try:
        uport = int(config.get('main', 'unicast_port'))
        DebugMessage("Using %d as unicast port" % int(uport))
    except ConfigParser.NoOptionError:
        uport = 5060
    try:
        conf_local_ip = config.get('main', 'local_ip')
        if is_valid_ipv4_address(conf_local_ip):
            local_ip = conf_local_ip
            DebugMessage("local_ip: %s" % local_ip)

        elif conf_local_ip == "default": #if loca_if == "default" try to reach snom.com
            try:
                local_ip = get_ip_address()
            except Exception, e:
                ConfigError("cannot discover local ip address: %s" % e)
                sys.exit(1)

        else: #network interface name
            try:
                local_ip = get_ip_address_ifname(conf_local_ip)
            except Exception, e:
                ConfigError("cannot determine ip address of %s interface: %s" % (conf_local_ip, e))
                sys.exit(1)

    except ConfigParser.NoOptionError:
        ConfigError("Missing mandatory parameters in config file, bailing out!")
        sys.exit(1)

    try:
        log_file = config.get('main', 'log_file')
        if log_file.startswith("syslog"):
            try:
                syslog_host = log_file.split(":")[1]
            except IndexError:
                syslog_host = 'localhost'
            try:
                syslog_port = int(log_file.split(":")[2])
            except IndexError:
                syslog_port = 514
            try:
                syslog_facility = log_file.split(":")[3]
            except IndexError:
                syslog_facility = logging.handlers.SysLogHandler.LOG_USER
            DebugMessage("Logging to syslog (host: %s, port: %s, facility: %s)" % ((syslog_host, syslog_port, syslog_facility)))
            conf_log_handler = logging.handlers.SysLogHandler((syslog_host, syslog_port), syslog_facility)
        else:
            DebugMessage("Logging to file: %s" % log_file)
            conf_log_handler = logging.FileHandler(log_file)

        conf_log_handler.setFormatter(formatter)
        logger.removeHandler(log_handler)
        logger.addHandler(conf_log_handler)

        InfoMessage("New server started")
    except ConfigParser.NoOptionError:
        # no log defined in config file
        pass

    regex = []
    subst = []
    run_commands = []

    try:
        rules = sorted([i for i in config.sections() if i.startswith("rule_")], key=lambda num: int(num.split("_")[1]))
    except ValueError, e:
        ConfigError("Invalid rule name: rule name mast be in format rule_XX, where XX is a number, eg. rule_13")
        ConfigError("Config details: %s" % e)
        sys.exit(1)

    InfoMessage("Valuating config sections order: %s" % rules)

    for s in rules:
        if s.startswith("rule_"):
            DebugMessage("Rule: %s" % s)
            try:
                reg = config.get(s, 'regex')
                sub = config.get(s, 'subst')
            except ConfigParser.NoOptionError:
                ConfigError("Missing regexp/subst option in %s section" % s)
                sys.exit(1)
            DebugMessage("Regex: %s" % reg)
            ###DebugMessage("Subst: %s" % sub)
            try:
                command = config.get(s, 'command')
                DebugMessage("Command: %s" % command)
            except ConfigParser.NoOptionError:
                command = None

            regex.append(ur'%s' % reg)
            subst.append(ur'%s' % sub)
            run_commands.append(command)

    server = MCastServer(local_ip=local_ip, regex=regex, subst=subst, run_commands=run_commands, uport=uport)

    if sys.argv[1] == "-c":
        request = "".join(sys.stdin.readlines())
        print "=== Request ==="
        print request
        resp = server.check_request(request)
        if resp:
            print "=== Response NOTIFY body ==="
            print resp
            sys.exit()
        else:
            print "No rules found"
            sys.exit(255)

    if sys.argv[1] == "-s":
        try:
            if config.get('main', 'daemon').upper() == 'TRUE':
                InfoMessage('Daemonizing')
                try:
                    pid_file = config.get('main', 'pid_file')
                    InfoMessage('Using pid file %s' % pid_file)
                    try:
                        pid = daemon.become_daemon(pid_file)
                    except Exception, e:
                        SystemError("Cannot start daemon: %s, exiting" %e)
                        sys.exit(-1)
                except ConfigParser.NoOptionError:
                    try:
                        pid = daemon.become_daemon(None)
                    except Exception, e:
                        SystemError("Cannot start daemon: %s, exiting" %e)
                        sys.exit(-1)
                    InfoMessage('No pid file in configuration file')
                InfoMessage("Daemon started with pid %d" % pid)

        except ConfigParser.NoOptionError:
            InfoMessage('Run in foreground')

        server.listen()
